lib/base/utility.hpp

```diff
@@ -157,6 +174,72 @@ class Utility
 	static boost::thread_specific_ptr<unsigned int> m_RandSeed;
 };
 
+#ifndef _WIN32
+template<size_t exceptAmount>
+inline void Utility::CloseAllFDs(const std::array<int, exceptAmount>& except)
+{
+	CloseAllFDs(except, [](int) {});
+}
+
+template<size_t exceptAmount, class OnClose>
+inline void Utility::CloseAllFDs(const std::array<int, exceptAmount>& except, OnClose onClose)
+{
+#if defined(__linux__) || defined(__APPLE__)
+	namespace fs = boost::filesystem;
+
+	std::set<int> fds;
+
+	for (fs::directory_iterator current (fs::path(
+#ifdef __linux__
+		"/proc/self/fd"
+#endif /* __linux__ */
+#ifdef __APPLE__
+		"/dev/fd"
+#endif /* __APPLE__ */
+	)), end; current != end; ++current) {
+		auto entry (current->path().filename());
+		int fd;
+
+		try {
+			fd = boost::lexical_cast<int>(entry.c_str());
+		} catch (...) {
+			continue;
+		}
+
+		fds.emplace(fd);
+	}
+
+	for (auto fd : except) {
+		fds.erase(fd);
+	}
+
+	for (auto fd : fds) {
+		if (close(fd) >= 0) {
+			onClose(fd);
+		}
+	}
+#else /* __linux__ || __APPLE__ */
+	rlimit rl;
+
+	if (getrlimit(RLIMIT_NOFILE, &rl) >= 0) {
+		rlim_t maxfds = rl.rlim_max;
+
+		if (maxfds == RLIM_INFINITY) {
+			maxfds = 65536;
+		}
+
+		std::set<int> fds (except.begin(), except.end());
```

Comment:
As `except` is rather small typically (has at most 4 elements right now), this won't bring any performance improvement. I wouldn't be surprised if this was even slower.

Reply:
O(3) slower than O(4)? C'mon.

Reply:
In the end, it's not about performance. Every single `close` syscall is probably more expensive and this is probably only done after `fork` which is a quite expensive syscall. So my reasoning is that this is line is just pointless as using `except.find(fd)` below would work just as fine.

Reply:
Why is it pointless if it’s faster?

Reply:
It's not. By the way, it's using `malloc` internally, so I think you would want to avoid it. ;)

Reply:
1. Why isn’t O(3) faster than O(4)?
2. But only on non-Linux and non-OSX \*nixes.

Reply:
Asymptotic complexity doesn't help that much when considering actual execution times for small input sizes. There, stuff like implementation details and caching have a much higher impact. For a std::vector, you have a contiguous piece of memory, that will fit in 1 or 2 cache lines whereas with a set, you will have a tree scattered around memory, where you have to dereference pointers all the time. See also for example the implementation of `std::sort` und GNU libstdc++ where they use insertion sort in some cases, even though this is O(n²) (https://en.wikipedia.org/wiki/Sort_(C%2B%2B)#Complexity_and_implementations).

See also the following example benchmark, `std::find` on a `std::vector` is about 65% faster than `std::set::find` (with both containing 4 elements) on my machine when compiling the following with `g++ -O2 -std=c++11`:
<details>

```cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <set>
#include <vector>

int main() {
	std::vector<int> v = {23, 42, 1337, 4711};
	std::set<int> s(v.begin(), v.end());

	for (int iter = 1; iter < (1<<30); iter <<= 1) {
		int set_count = 0, vec_count = 0;
		int64_t set_ns, vec_ns;
		{
			auto start = std::chrono::high_resolution_clock::now();
			for (int i = 0; i < iter; i++) {
				set_count += s.find(i) != s.end();
			}
			auto finish = std::chrono::high_resolution_clock::now();
			set_ns = std::chrono::nanoseconds(finish - start).count();
			std::cout << "set    " << iter << " iterations: " << set_ns << "ns (" << set_count << ")" << std::endl;
		}

		{
			auto start = std::chrono::high_resolution_clock::now();
			for (int i = 0; i < iter; i++) {
				vec_count += std::find(v.begin(), v.end(), i) != v.end();
			}
			auto finish = std::chrono::high_resolution_clock::now();
			vec_ns = std::chrono::nanoseconds(finish - start).count();
			std::cout << "vector " << iter << " iterations: " << vec_ns << "ns (" << vec_count << ")" << std::endl;
		}

		std::cout << "delta: " << (vec_ns - set_ns) << "ns, ratio: " << (double(vec_ns) / set_ns) << std::endl;
	}
}
```
</details>